src/security/Handshake.cc

```diff
@@ -504,6 +521,78 @@ Security::HandshakeParser::parseSniExtension(const SBuf &extensionData) const
     return SBuf(); // SNI extension lacks host_name
 }
 
+/// RFC 8446 Section 4.2.1: SupportedVersions extension
+void
+Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extensionData) const
+{
+    // Upon detecting a quoted RFC MUST violation, this parser immediately
+    // returns, ignoring the entire extension and resulting in Squid relying on
+    // the legacy_version field value or another (valid) supported_versions
+    // extension. The alternative would be to reject the whole handshake as
+    // invalid. Deployment experience will show which alternative is the best.
+
+    // Please note that several of these MUSTs also imply certain likely
+    // handling of a hypothetical next TLS version (e.g., v1.4).
+
+    // RFC 8446 Section 4.1.2:
+    // In TLS 1.3, the client indicates its version preferences in the
+    // "supported_versions" extension (Section 4.2.1) and the legacy_version
+    // field MUST be set to 0x0303, which is the version number for TLS 1.2.
+    //
+    // RFC 8446 Section 4.2.1:
+    // A server which negotiates TLS 1.3 MUST respond by sending a
+    // "supported_versions" extension containing the selected version value
+    // (0x0304).  It MUST set the ServerHello.legacy_version field to 0x0303
+    // (TLS 1.2).
+    //
+    // Ignore supported_versions senders violating legacy_version MUSTs above:
+    if (details->tlsSupportedVersion != AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, 2))
```

Comment:
To reduce cycles initializing stack memory this version object should be a local constant. 
```suggestion
    static const AnyP::ProtocolVersion Tls1p2(AnyP::PROTO_TLS, 1, 2);
    if (details->tlsSupportedVersion != Tls1p2)
```

Reply:
> To reduce cycles initializing stack memory this version object should be a local constant.

I am sorry, but the above suggestion makes things _slower_ rather than faster when it comes to modern optimizing compilers. The suggestion is based on several incorrect assumptions as detailed further below. In short, adding this local static variable will often make performance slightly worse because the ProtocolVersion class is designed, in part, to _inline_ comparisons with constant/known versions. In this and similar contexts, inlining is much faster than using a local static.

FYI: A local static variable (including constants) requires a "have I been initialized?" check _every time_ the local code is executed. That check usually requires a mutex-protected(!) access that (AFAICT) forces even modern compilers to turn off many valuable optimizations. Local statics are a valuable tool for reducing creation overhead of expensive-to-create objects in non-threaded code, but they are not appropriate for replacing fully inlined comparisons with trivial compile-time constants.

@yadij, please withdraw this change request (and another similar one). If you would like, we can add a `TlsV(x,y)` wrapper to make inlined calls like this visually shorter.

----

P.S. I am sure you will just trust that I know what I am saying, but in case somebody else reads this and have doubts, they can confirm my conclusions by comparing fooTemp() and fooStatic() assembly [code](https://godbolt.org/z/QRchPf) designed to mimic the two implementations of ProtocolVersion comparison discussed above. fooTemp() is ~10 quick assembly instructions while fooStatic() needs more than 40, and even the essential/similar segments of fooStatic() are slower.

BTW, either removing `static` or moving the static constant outside the function allows at least some modern compilers to optimize the suggested variable away -- fooStatic() compiles into the same code as fooTemp()! This disappearance helps demonstrate the cost of local statics. Should we adjust the suggestion to remove `static` or make the constant global? No, we should still prefer the currently implemented version because it is stateless.